README.md


JavaScript client for [Sematext Logs](http://sematext.com/logsene).

# Installation

    npm install logsene-js --save

# Usage


    var Logsene = require('logsene-js')
    var logger =  new Logsene ('LOGS-APP-TOKEN')
    logger.log ('info', 'text message', {tags: ['a','b'], customField: 'custom-field'})

### Constructor Parameters

- __token__ - Logs App token. Sign up [here](https://apps.sematext.com/users-web/register.do).
- __type__ - Optional. Default type of your logs - please note you can define [Elasticsearch mapping templates in Sematext Logs](http://blog.sematext.com/elasticsearch-mapping-types-for-json-logging/)
- __url__ - Sematext Logs receiver URL (e.g. if you are shipping logs to [Sematext Enterprise](https://sematext.com/enterprise)), defaults to ```'https://logsene-receiver.sematext.com/_bulk'```
- __storageDirectory__ - Directory where to buffer logs in the case of failure
- __options__: 
  - __useIndexInBulkUrl__ -  If set to 'false' /_bulk will be used /indexName/_bulk otherwise.
  - __httpOptions__ - general HTTP/HTTPS options for the [request](https://nodejs.org/api/https.html#https_https_request_options_callback), e.g. SSL key,cert,passphrase,ca,rejectUNauthorized etc. 
  - __silent__ - If set to `true`, logsene-js will not log debug and errors to stdout. Used for prod envs, but not mandatory. This is `false` by default.

## Special fields for indexing

In general Elasticsearch > 2.3 (including Sematext Logs) does not allow fields with leading underscore or dots in field names. Logsene-js converts such fields names (e.g. removing leading underscores, and replaces dots to underscores). However, a few fields are interpreted for indexing before renaming the fields: 
- _type - used as '_type' in the index operation (bulk indexing)
- _id - used as '_id' in the index operation (bulk indexing)


## Environment variables
- LOGS_TMP_DIR - Directory to store failed bulk requests, for later re-transmission. Failed requests are not stored, when LOGS_TMP_DIR is not set.
- LOG_INTERVAL - Time to collect logs before a bulk request is done. Default 10000 ms
- LOGS_BULK_SIZE - Maximum size of a bulk request. Default 1000.
- LOGS_MAX_MESSAGE_FIELD_SIZE - maximum size of the 'message' field, default 240k
- LOGS_RECEIVER_URL - URL for the Sematext Logs receiver. E.g. for Sematext Enterprise version. Defaults to Sematext Logs SaaS receiver https://logsene-receiver.sematext.com/_bulk
- LOGSENE_TMP_DIR - Directory where to buffer logs if the Elasticsearch endpoint is unavailable. If this is not set, `logsene-js` will not buffer logs at all.
- LOGS_BUFFER_ON_APP_LIMIT - default 'true'. HTTP bulk requests are retried until Sematext starts accepting logs again. Please increase your daily limit setting in Logs App settings if you see "403, App limit reached" errors. Setting the value to 'false' will disable disk buffering only for failed http requests '403, Forbidden / App limit reached'. We recommend to keep the default setting 'true' to avoid any loss of logs. 
- LOGS_REMOVE_FIELDS - a comma separated list of fields, which should not be logged. For nested objects use a dot notation e.g. 'reques.body,request.size'

__Note:__ Previous versions of logsene-js used LOGSENE instead of LOGS prefix for the settings above. Logsene-js is backward compatible to previous environment variable names. However all variable names with the LOGSENE prefix are depracated and might be removed in future relases.

## Security

- HTTPS is enabled by default
- Environment variables for Proxy servers:
  - For HTTPS endpoints (default): HTTPS_PROXY / https_proxy
```
        export HTTPS_PROXY=https://my-ssl-proxy.example
        export HTTPS_PROXY=http://my-proxy.example
```
  - For HTTP endpoints (e.g. On-Premises): HTTP_PROXY / http_proxy
```
        export HTTP_PROXY=http://my-proxy.example
        export HTTP_PROXY=https://my-ssl-proxy.example
```

# Other related modules

- Please check [winston-logsene](https://github.com/sematext/winston-logsene) a transport layer for the winston logging framework.
- Please see [bunyan-logsene](https://github.com/6RiverSystems/bunyan-logsene) a transport layer for the bunyan logging framework.


CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

#### [1.1.77](https://github.com/sematext/logsene-js/compare/1.1.76...1.1.77)

- Fix node fetch issue and vulnerabilities [`#55`](https://github.com/sematext/logsene-js/pull/55)
- Housekeeping [`#51`](https://github.com/sematext/logsene-js/pull/51)
- chore: remove unused libraries and code [`#50`](https://github.com/sematext/logsene-js/pull/50)

#### [1.1.76](https://github.com/sematext/logsene-js/compare/1.1.75...1.1.76)

> 20 April 2023

- Sc 15901 remove vulnerabilities [`#49`](https://github.com/sematext/logsene-js/pull/49)
- Bump vm2 from 3.9.15 to 3.9.16 [`#47`](https://github.com/sematext/logsene-js/pull/47)
- Bump http-cache-semantics from 4.1.0 to 4.1.1 [`#43`](https://github.com/sematext/logsene-js/pull/43)
- Bump vm2 from 3.9.12 to 3.9.15 [`#45`](https://github.com/sematext/logsene-js/pull/45)
- Bump shelljs from 0.8.4 to 0.8.5 [`#41`](https://github.com/sematext/logsene-js/pull/41)
- Bump json-schema and jsprim [`#39`](https://github.com/sematext/logsene-js/pull/39)
- Bump decode-uri-component from 0.2.0 to 0.2.2 [`#35`](https://github.com/sematext/logsene-js/pull/35)
- Bump ansi-regex from 3.0.0 to 3.0.1 [`#36`](https://github.com/sematext/logsene-js/pull/36)
- Bump minimatch and mocha [`#37`](https://github.com/sematext/logsene-js/pull/37)
- Bump qs from 6.5.2 to 6.5.3 [`#34`](https://github.com/sematext/logsene-js/pull/34)
- Bump minimist from 1.2.5 to 1.2.6 [`#32`](https://github.com/sematext/logsene-js/pull/32)
- Bump node-fetch from 2.6.1 to 2.6.7 [`#31`](https://github.com/sematext/logsene-js/pull/31)
- Bump ajv from 6.12.2 to 6.12.6 [`#30`](https://github.com/sematext/logsene-js/pull/30)
- Bump path-parse from 1.0.6 to 1.0.7 [`#28`](https://github.com/sematext/logsene-js/pull/28)
- Bump glob-parent from 5.1.1 to 5.1.2 [`#27`](https://github.com/sematext/logsene-js/pull/27)
- chore: remove unused libraries and code [`b76d6a8`](https://github.com/sematext/logsene-js/commit/b76d6a85746e115635b17fbfd5b07516f4362199)
- Release 1.1.76 [`4c9aa4c`](https://github.com/sematext/logsene-js/commit/4c9aa4c67443a67c2ed71959a1fd58ea8f5129f0)
- Fixed user strict typo [`1583aae`](https://github.com/sematext/logsene-js/commit/1583aae5e7ca7498e1b3db849bc0335be6b13dde)

#### [1.1.75](https://github.com/sematext/logsene-js/compare/1.1.74...1.1.75)

> 2 June 2021

- Bump hosted-git-info from 2.8.8 to 2.8.9 [`#26`](https://github.com/sematext/logsene-js/pull/26)
- Bump y18n from 4.0.0 to 4.0.1 [`#25`](https://github.com/sematext/logsene-js/pull/25)
- Bump ini from 1.3.5 to 1.3.8 [`#21`](https://github.com/sematext/logsene-js/pull/21)
- updated deps [`953154d`](https://github.com/sematext/logsene-js/commit/953154d1b36f532e1476c1856658e4463cf976cc)
- Release 1.1.75 [`0e87c99`](https://github.com/sematext/logsene-js/commit/0e87c996003e33a139a19c4c95e5a86ab5988e44)

#### [1.1.74](https://github.com/sematext/logsene-js/compare/1.1.73...1.1.74)

> 8 October 2020

- added logType [`56978cc`](https://github.com/sematext/logsene-js/commit/56978cc4dc83b93809ffcc84ab992e171eae4631)
- Release 1.1.74 [`80c5cfb`](https://github.com/sematext/logsene-js/commit/80c5cfb1fb981f368bb1646dedd106da05706b28)

#### [1.1.73](https://github.com/sematext/logsene-js/compare/1.1.72...1.1.73)

> 23 September 2020

- Bump node-fetch from 2.6.0 to 2.6.1 [`#18`](https://github.com/sematext/logsene-js/pull/18)
- Release 1.1.73 [`922d98a`](https://github.com/sematext/logsene-js/commit/922d98a2f391735a1522b393d6ee0e835b161d63)

#### [1.1.72](https://github.com/sematext/logsene-js/compare/1.1.71...1.1.72)

> 20 August 2020

- edited test cases to use new error handling logic [`376cde5`](https://github.com/sematext/logsene-js/commit/376cde57bf38b10add67775a0a9bc032aa88ecd6)
- optimize memory limit test [`a1d7275`](https://github.com/sematext/logsene-js/commit/a1d7275fe1afd527501b4ae5230f6c8c5ce51401)
- Release 1.1.72 [`62a2cd7`](https://github.com/sematext/logsene-js/commit/62a2cd76134bb969081b3fbac7264ca5619d0a28)

#### [1.1.71](https://github.com/sematext/logsene-js/compare/1.1.70...1.1.71)

> 7 August 2020

- added better error handling [`ba45c88`](https://github.com/sematext/logsene-js/commit/ba45c881bfb42823acfb0ffb22141fcc18c29f9a)
- Release 1.1.71 [`c5c00dc`](https://github.com/sematext/logsene-js/commit/c5c00dcb82894cb0354afe78b6d9afb30a3e6015)

#### [1.1.70](https://github.com/sematext/logsene-js/compare/1.1.69...1.1.70)

> 6 August 2020

- Release 1.1.70 [`4c6c472`](https://github.com/sematext/logsene-js/commit/4c6c47234a617431793894714a67bf292090fd2f)
- added npm config [`1fa4545`](https://github.com/sematext/logsene-js/commit/1fa4545d9d5694ab66fc4dd26d171e82a8b33221)

#### [1.1.69](https://github.com/sematext/logsene-js/compare/1.1.68...1.1.69)

> 6 August 2020

- Typescript definitions [`#16`](https://github.com/sematext/logsene-js/pull/16)
- add changelog [`ca08b5a`](https://github.com/sematext/logsene-js/commit/ca08b5a629ef7615ee58243dc86215947097c598)
- audited deps [`d306835`](https://github.com/sematext/logsene-js/commit/d3068357a9dae26954261689efeda50e5bbae7b0)
- removed auto changelog dev dep [`16ce84f`](https://github.com/sematext/logsene-js/commit/16ce84f18c99c8a68a4bd893c2e4c47d8747a7b4)

#### [1.1.68](https://github.com/sematext/logsene-js/compare/1.1.67...1.1.68)

> 28 March 2020


index.d.ts
type LogseneHttpOptions = {
  maxSockets?: number;
  keepAlive?: boolean;
  maxFreeSockets?: number;
  // Allow for additional options that might not be accounted for
  [key: string]: any;
};

type LogseneOptions = {
  useIndexInBulkUrl?: boolean;
  httpOptions?: {[key: string]: any};
  // Allow for additional options that might not be accounted for
  [key: string]: any;
};

type LogFields = {
  _type?: string;
  _id?: string;
  [key: string]: any;
};

type Callback = (err: object, msg: object) => void;

declare class Logsene {
  constructor(
    token: string,
    type?: string,
    url?: string,
    storageDirectory?: string,
    options?: LogseneOptions
  );

  log(
    level: string,
    message: string,
    fields?: LogFields,
    callback?: Callback
  ): void;

  send(callback: Callback): void;
}

export = Logsene;

index.js
/*
 * See the NOTICE.txt file distributed with this work for additional information
 * regarding copyright ownership.
 * Sematext licenses logsene-js to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

'use strict'
const nodeFetch = require('node-fetch')
const fs = require('fs')
const util = require('util')
const os = require('os')
const events = require('events')
const ipAddress = require('ip').address()
const path = require('path')
const stringifySafe = require('fast-safe-stringify')
const streamBuffers = require('stream-buffers')

// settings for node stream buffer
const initialBufferSize = 1024 * 1024
const incrementBuffer = 1024 * 1024
// re-usable regular expressions
const limitRegex = /limit/i
const appNotFoundRegEx = /Application not found for token/i
const disableJsonEnrichment = (process.env.ENABLE_JSON_ENRICHMENT === 'false')
let replaceDotsInFieldNames = true
if (process.env.REPLACE_DOTS_IN_FIELD_NAMES === 'false' || process.env.REPLACE_DOTS_IN_FIELD_NAMES === '0') {
  replaceDotsInFieldNames = false
}
// load ENV like Logsene receivers from file containing
// env vars e.g. SPM_RECEIVER_URL, EVENTS_RECEIVER_URL, LOGSENE_RECEIVER_URL
// the file overwrites the actual environment
// and is used by Sematext Enterprise or multi-region setups to
// setup receiver URLs
function loadEnvFromFile (fileName) {
  try {
    const receivers = fs.readFileSync(fileName).toString()
    if (receivers) {
      const lines = receivers.split('\n')
      lines.forEach(function (line) {
        const kv = line.split('=')
        if (kv.length === 2 && kv[1].length > 0) {
          process.env[kv[0].trim()] = kv[1].trim()
          if (/logsene-js/.test(process.env.DEBUG)) {
            console.log(kv[0].trim() + ' = ' + kv[1].trim())
          }
        }
      })
    }
    if (/logsene-js/.test(process.env.DEBUG)) {
      console.log(new Date(), 'loading Sematext receiver URLs from ' + fileName)
    }
  } catch (error) {
    // ignore missing file or wrong format
    if (/logsene-js/.test(process.env.DEBUG)) {
      console.error(error.message)
    }
  }
}
const envFileName = '/etc/sematext/receivers.config'
/**
  if (/win/.test(os.platform()) {
    envFileName = process.env.ProgramData + '\\Sematext\\receivers.config'
  }
**/
loadEnvFromFile(envFileName)

// removing LOGSENE from ENV variable names
// and be backward compatible in case users still use old ENV variable names
const envVarMapping = [
  ['LOGSENE_MAX_MESSAGE_FIELD_SIZE', 'LOGS_MAX_MESSAGE_FIELD_SIZE'],
  ['LOGSENE_MAX_STORED_REQUESTS', 'LOGS_MAX_STORED_REQUESTS'],
  ['LOGSENE_BULK_SIZE_BYTES', 'LOGS_BULK_SIZE_BYTES'],
  ['LOGSENE_BULK_SIZE', 'LOGS_BULK_SIZE'],
  ['LOGSENE_LOG_INTERVAL', 'LOG_INTERVAL'],
  ['LOGSENE_DISK_BUFFER_INTERVAL', 'LOGS_DISK_BUFFER_INTERVAL'],
  ['LOGSENE_RECEIVER_URL', 'LOGS_RECEIVER_URL'],
  ['SPM_RECEIVER_URL', 'MONITORING_RECEIVER_URL'],
  ['SPM_TOKEN', 'MONITORING_TOKEN'],
  ['LOGSENE_REMOVE_FIELDS', 'REMOVE_FIELDS'],
  ['LOGSENE_TMP_DIR', 'LOGS_TMP_DIR'],
  ['LOGSENE_BUFFER_ON_APP_LIMIT ', 'LOGS_BUFFER_ON_APP_LIMIT'],
  ['LOGSENE_REMOVE_FIELDS', 'LOGS_REMOVE_FIELDS'],
  ['LOGSENE_REMOVE_FIELDS', 'REMOVE_FIELDS'],
  ['SPM_REPORTED_HOSTNAME', 'REPORTED_HOSTNAME']
]

function mapEnv (item) {
  if ((!process.env[item[0]]) && (process.env[item[1]] !== undefined)) {
    process.env[item[0]] = process.env[item[1]]
    if (/logsene-js/.test(process.env.DEBUG)) {
      console.log('Set ', item[0], ' to ', process.env[item[0]], ' from ', item[1])
    }
  }
}
envVarMapping.forEach(mapEnv)

// SPM_REPORTED_HOSTNAME might be set by Sematext Docker Agent
// the container hostname might not be helpful ...
// this might be removed after next release of SDA setting xLogseneOrigin from SDA
var xLogseneOrigin = process.env.SPM_REPORTED_HOSTNAME || os.hostname()
// limit message size
var MAX_MESSAGE_FIELD_SIZE = Number(process.env.LOGSENE_MAX_MESSAGE_FIELD_SIZE) || 1024 * 240 // 240 K, leave
// settings for bulk requests
var MIN_LOGSENE_BULK_SIZE = 200
var MAX_LOGSENE_BULK_SIZE = 10000
var MAX_STORED_REQUESTS = Number(process.env.LOGSENE_MAX_STORED_REQUESTS) || 10000
var MAX_CLIENT_SOCKETS = Number(process.env.MAX_CLIENT_SOCKETS) || 5

// upper limit a user could set - 10 MB as configured in Sematext Cloud receivers
var MAX_LOGSENE_BULK_SIZE_BYTES = 10 * 1000 * 1000
// lower limit a user could set
var MIN_LOGSENE_BULK_SIZE_BYTES = 1024 * 1024
var MAX_LOGSENE_BUFFER_SIZE = Number(process.env.LOGSENE_BULK_SIZE_BYTES) || 1024 * 1024 * 3 // max 3 MB per http request
// check limits set by users, and adjust if those would lead to problematic settings
if (MAX_LOGSENE_BUFFER_SIZE > MAX_LOGSENE_BULK_SIZE_BYTES) {
  MAX_LOGSENE_BUFFER_SIZE = MAX_LOGSENE_BULK_SIZE_BYTES
}
if (MAX_LOGSENE_BUFFER_SIZE < MIN_LOGSENE_BULK_SIZE_BYTES) {
  MAX_LOGSENE_BUFFER_SIZE = MIN_LOGSENE_BULK_SIZE_BYTES
}
var LOGSENE_BULK_SIZE = Number(process.env.LOGSENE_BULK_SIZE) || 500 // max 500 messages per bulk req.
if (LOGSENE_BULK_SIZE > MAX_LOGSENE_BULK_SIZE) {
  LOGSENE_BULK_SIZE = MAX_LOGSENE_BULK_SIZE
}
if (LOGSENE_BULK_SIZE < MIN_LOGSENE_BULK_SIZE) {
  LOGSENE_BULK_SIZE = MIN_LOGSENE_BULK_SIZE
}

const deleteKey = require('del-key')

function removeFields (fieldList, doc) {
  if (fieldList && fieldList.length > 0 && fieldList[0] !== '') {
    for (let i = fieldList.length; i >= 0; --i) {
      deleteKey(doc, fieldList[i])
    }
  }
  return doc
}

// Create a deep clone of an object while allowing caller to rename
// keys, replace values, or reject key-pairs entirely.
//
// Does not modify the source object. Callback receives (key, value)
// and is expected to return a two-item array [newKey, newValue], or
// null if the pair should be absent from the resulting object.

function deepConvert (src, cb) {
  let dest
  if (Array.isArray(src)) {
    dest = []
  } else {
    dest = {}
  }
  if (dest) {
    for (let key in src) {
      if (src.hasOwnProperty(key)) {
        const val = src[key]
        const newKV = cb(key, val)
        if (newKV === null) {
          // skip this field entirely
          continue
        }

        const newKey = newKV[0]
        const newVal = newKV[1]

        if (newVal !== undefined &&
          newVal !== null &&
          (Array.isArray(newVal) || newVal.constructor === Object)) {
          dest[newKey] = deepConvert(newVal, cb)
        } else {
          dest[newKey] = newVal
        }
      }
    }
  } else {
    dest = src
  }

  return dest
}

/**
 * token - the LOGSENE Token
 * type - type of log (string)
 * url - optional alternative URL for Logsene receiver (e.g. for on premises version)
 * storageDirectory - directory where to buffer logs
 * options - @object { useIndexInBulkUrl, httpOptions, silent }
 */
function Logsene (token, type, url, storageDirectory, options) {
  if (!token) {
    throw new Error('Logsene token not specified')
  }
  if (options) {
    this.options = options
  } else {
    this.options = {
      useIndexInBulkUrl: false,
      silent: false
    }
  }
  if (url && /logsene/.test(url)) {
    // logs to logsene should use /TOKEN/_bulk
    this.options.useIndexInBulkUrl = true
  }
  this.request = null
  this.maxMessageFieldSize = MAX_MESSAGE_FIELD_SIZE
  this.xLogseneOrigin = xLogseneOrigin
  this.token = token
  this.setUrl(url || process.env.LOGSENE_URL || process.env.LOGSENE_RECEIVER_URL || process.env.LOGS_URL || 'https://logsene-receiver.sematext.com/_bulk')
  this.type = type || 'logs'
  this.hostname = process.env.SPM_REPORTED_HOSTNAME || os.hostname()
  this.bulkReq = new streamBuffers.WritableStreamBuffer({
    initialSize: initialBufferSize,
    incrementAmount: incrementBuffer
  })
  this.logCount = 0
  this.sourceName = null
  if (process.mainModule && process.mainModule.filename) {
    this.sourceName = path.basename(process.mainModule.filename)
  }
  events.EventEmitter.call(this)
  let self = this
  self.lastSend = Date.now()
  const logInterval = Number(process.env.LOGSENE_LOG_INTERVAL) || 20000
  const tid = setInterval(function () {
    if (self.logCount > 0 && (Date.now() - self.lastSend) > (logInterval - 1000)) {
      self.send()
    }
  }, logInterval)
  if (tid.unref) {
    tid.unref()
  }
  process.on('beforeExit', function () {
    self.send()
  })
  if (process.env.LOGSENE_TMP_DIR || storageDirectory) {
    this.diskBuffer(true, process.env.LOGSENE_TMP_DIR || storageDirectory)
  }
  const fieldListStr = process.env.LOGSENE_REMOVE_FIELDS || process.env.REMOVE_FIELDS || ''
  this.removeFieldsList = fieldListStr.replace(/ /g, '').split(',')

  // error handling
  self.on('x-logsene-error', function (errObj) {
    if (!self.options.silent) {
      console.error(
        new Date().toISOString() + 
        ` logsene-js: cannot reach the receiver URL: ${errObj.err.url}, please check the error message ...`,
        errObj
      )
    }
  })
}
util.inherits(Logsene, events.EventEmitter)

Logsene.prototype.setUrl = function (url) {
  let tmpUrl = url
  if (url.indexOf('_bulk') === -1) {
    tmpUrl = url + '/_bulk'
  } else {
    tmpUrl = url
  }
  if (this.options && this.options.useIndexInBulkUrl) {
    this.url = tmpUrl.replace('_bulk', this.token + '/_bulk')
  } else {
    this.url = tmpUrl
  }
  let Agent = null
  let httpOptions = { maxSockets: MAX_CLIENT_SOCKETS, keepAlive: false, maxFreeSockets: MAX_CLIENT_SOCKETS }
  if (this.options.httpOptions) {
    const keys = Object.keys(this.options.httpOptions)
    for (let i = 0; i < keys.length; i++) {
      httpOptions[keys[i]] = this.options.httpOptions[keys[i]]
    }
  }
  if (/^https/.test(url)) {
    Agent = require('https').Agent
  } else {
    Agent = require('http').Agent
  }
  this.httpAgent = new Agent(httpOptions)
  this.httpDefaults = {
    agent: this.httpAgent,
    timeout: 60000,
  }
}
const DiskBuffer = require('./DiskBuffer.js')

Logsene.prototype.diskBuffer = function (enabled, dir) {
  if (enabled) {
    const tmpDir = path.join((dir || require('os').tmpdir()), this.token)
    this.db = DiskBuffer.createDiskBuffer({
      tmpDir: tmpDir,
      maxStoredRequests: Number(MAX_STORED_REQUESTS),
      interval: process.env.LOGSENE_DISK_BUFFER_INTERVAL || 5000
    })
    this.db.syncFileListFromDir()
    let self = this
    if (!this.db.isCached) {
      // only the first instance registers for retransmit-req
      // to avoid double event handling from multiple instances
      this.db.on('retransmit-req', function (event) {
        self.shipFile(event.fileName, event.buffer, function (err, res) {
          if (err && err.httpBody && appNotFoundRegEx.test(err.httpBody)) {
            // remove file from DiskBuffer when token is invalid
            self.db.rmFile.call(self.db, event.fileName)
            self.db.retransmitNext.call(self.db)
            return
          }
          if (!err && res) {
            self.db.rmFile.call(self.db, event.fileName)
            self.db.retransmitNext.call(self.db)
          } else {
            self.db.unlock.call(self.db, event.fileName)
          }
        })
      })
    }
  }
  this.persistence = enabled
}

/**
 * Add log message to send buffer
 * @param level - log level e.g. 'info', 'warning', 'error'
 * @param message - text message
 * @param fields - Object with custom fields or overwrite of any other field e.g. e.g. "{@timestamp: new Date.toISOString()}"
 * @param callback (err, msg object)
 */
Logsene.prototype.log = function (level, message, fields, callback) {
  this.logCount = this.logCount + 1
  let type = 'logs'
  let logType = 'logs'

  if (fields) {
    logType = fields._type
  }

  if (this.options.useIndexInBulkUrl) {
    // not a Sematext service -> use only one type per index
    // Elasticsearch > 6.x allows only one type per index
    type = this.type
  }

  let msg = {
    '@timestamp': new Date(),
    severity: level,
    message,
    logType,
    os: {
      host: this.hostname,
      hostip: ipAddress
    }
  }

  let elasticsearchDocId = null
  if (fields && fields._type) {
    delete fields._type
  }
  if (fields && fields._id) {
    elasticsearchDocId = fields._id
  }

  if (disableJsonEnrichment) {
    msg = {}
  }
  let esSanitizedFields = deepConvert(fields, function (key, val) {
    if (typeof val === 'function') {
      return null
    } else {
      if (replaceDotsInFieldNames) {
        return [key.replace(/\./g, '_').replace(/^_+/, ''), val]
      } else {
        // remove leading _
        return [key.replace(/^_+/, ''), val]
      }
    }
  })
  msg = Object.assign(msg, esSanitizedFields)
  if (msg['@timestamp'] && typeof msg['@timestamp'] === 'number') {
    msg['@timestamp'] = new Date(msg['@timestamp'])
  }
  msg = removeFields(this.removeFieldsList, msg)
  let _index = this.token
  if (fields && typeof (fields._index) === 'function') {
    _index = fields._index(msg)
  } else {
    if (fields && fields._index != undefined) {
      _index = fields._index
      delete msg.index
    }
  }

  if (msg.message && typeof msg.message === 'string' && Buffer.byteLength(msg.message, 'utf8') > this.maxMessageFieldSize) {
    // new nodejs api, let's use Buffer.alloc instead of Buffer(size)
    let cutMsg = Buffer.alloc ? Buffer.alloc(this.maxMessageFieldSize) : new Buffer(this.maxMessageFieldSize)
    cutMsg.write(msg.message)
    msg.message = cutMsg.toString()
    if (msg.originalLine) {
      // when message is too large and logagent added originalLine,
      // this should be removed to stay under the limits in receiver
      delete msg.originalLine
    }
    msg.logsene_client_warning = 'Warning: message field too large > ' + this.maxMessageFieldSize + ' bytes'
  }

  if (elasticsearchDocId !== null) {
    this.bulkReq.write(stringifySafe({ 'index': { '_index': _index, '_id': String(elasticsearchDocId), '_type': type || this.type } }) + '\n')
  } else {
    this.bulkReq.write(stringifySafe({ 'index': { '_index': _index, '_type': type || this.type } }) + '\n')
  }

  this.bulkReq.write(stringifySafe(msg) + '\n')

  if (this.logCount === LOGSENE_BULK_SIZE || this.bulkReq.size() > MAX_LOGSENE_BUFFER_SIZE) {
    this.send()
  }
  this.emit('logged', { msg: msg, _index: _index })
  if (callback) {
    callback(null, msg)
  }
}

/**
 * Sending log entry  to LOGSENE - this function is triggered every 100 log message or 30 seconds.
 * @callback {function} optional callback function
 */
Logsene.prototype.send = function (callback) {
  let self = this
  const buffer = this.bulkReq
  this.bulkReq = new streamBuffers.WritableStreamBuffer({
    initialSize: initialBufferSize,
    incrementAmount: incrementBuffer
  })
  buffer.end()
  self.lastSend = Date.now()
  let count = this.logCount
  let options = {
    url: this.url,
    logCount: count,
    headers: {
      'User-Agent': 'logsene-js',
      'Content-Type': 'application/json',
      'Connection': 'Close',
      'x-logsene-origin': this.xLogseneOrigin || xLogseneOrigin
    },
    body: buffer.getContents(),
    method: 'POST'
  }

  if (options.body === false) {
    return
  }
  function httpResult (err, res) {
    // if (res && res.body) console.log(res.statusCode, res.body)
    let logseneError = null
    let responseJson = null
    if (res && res.headers && res.headers['x-logsene-error']) {
      logseneError = res.headers['x-logsene-error']
    }
    let errorMessage = null
    let errObj = null
    if (err || (res && res.status > 399) || logseneError) {
      if (err && (err.code || err.message)) {
        err.url = options.url
      }
      if (res && res.status) {
        errorMessage = 'HTTP status code:' + res.status
      }

      if (logseneError) {
        errorMessage += ', ' + logseneError
      }
      errObj = { source: 'logsene-js', err: (err || { message: errorMessage, httpStatus: res.status, httpBody: res.body, url: options.url }) }
      self.emit('x-logsene-error', errObj)

      if (self.persistence) {
        let storeFileFlag = true
        // don't use disk buffer for invalid Logsene tokens
        if (res && res.body && appNotFoundRegEx.test(res.body)) {
          storeFileFlag = false
        }
        if (res && res.status && res.status === 400) {
          storeFileFlag = false
        }
        if (logseneError && limitRegex.test(logseneError)) {
          // && process.env.LOGSENE_BUFFER_ON_APP_LIMIT === 'false'
          storeFileFlag = false
        }
        if (storeFileFlag) {
          options.body = options.body.toString()
          self.db.store(options, function () {
            delete options.body
          })
        } else {
          self.emit('fileNotStored', options)
        }
      }
    } else {
      return res.json()
        .then(responseJson => {
          responseJson.items.forEach(function (item) {
            let result = item.index || item.create || item.update || item.delete
            if (result && result.status > 399) {
              errorMessage = 'HTTP status code:' + result.status + ' Error: ' + JSON.stringify(result.error)
              let errObj = {
                source: 'logsene-js',
                err: { message: errorMessage, httpStatus: result.status, httpBody: result, url: options.url }
              }
              self.emit('x-logsene-error', errObj)
            }
          })

          self.emit('log', { source: 'logsene-js', count: count, url: options.url })
          delete options.body
          if (callback) {
            callback(errObj, res)
          }
        })
        .catch(err => {
          errObj = { source: 'logsene-js', err: err }
          self.emit('x-logsene-error', errObj)
        })
    }
  }
  self.logCount = Math.max(self.logCount - count, 0)
  options = {...options, ...this.httpDefaults}
  nodeFetch(this.url, options)
    .then(response => httpResult(null, response))
    .catch(err => httpResult(err, null))
}

Logsene.prototype.shipFile = function (name, data, cb) {
  let self = this
  let options = null
  try {
    options = JSON.parse(data)
  } catch (err) {
    // wrong file format
  }
  if (options == null || options.options) {
    // wrong file format?
    // cleanup from earlier versions
    // self.db.rmFile(name)
    return cb(new Error('wrong bulk file format'))
  }
  options.body = options.body.toString()
  options.url = self.url
  const callbackFunc = function (err, res) {
    if (err || (res && res.status > 399)) {
      let errObj = { source: 'logsene re-transmit', err: (err || { message: 'Logsene re-transmit status code:' + res.status, httpStatus: res.status, httpBody: res.body, url: options.url, fileName: name }) }
      self.emit('x-logsene-error', errObj)

      if (cb) {
        cb(errObj)
      }
    } else {
      if (cb) {
        cb(null, { file: name, count: options.logCount })
      }
      self.emit('file shipped', { file: name, count: options.logCount })
      self.emit('rt', { count: options.logCount, source: 'logsene', file: name, url: String(options.url), request: null, response: null })
    }
  }
  nodeFetch(options.url, options)
    .then(response => callbackFunc(null, response))
    .catch(error => callbackFunc(error, null))
}
module.exports = Logsene
